File: mnist_in_30.py
import numpy as np
import mnist

def dim(a):
    if not type(a) == list:
        return []
    return [len(a)] + dim(a[0])
    
def feed_forward(X, weights):
    a = [X]
    for w in weights:
        a.append(np.maximum(a[-1].dot(w),0))
    return a

def grads(X, Y, weights):
    print("grads() called")
    grads = np.empty_like(weights)
    a = feed_forward(X, weights)
    #a[0].shape = (20,784)
    #a[1].shape = (20,100)
    #a[2].shape = (20,10)
    delta = a[-1] - Y #delta.shape: (20,10), Y.shape:(20,10)
    print(f"   a[-1].shape: {(a[-1]).shape}")
    print(f"   a[-2].shape: {(a[-2]).shape}")
    #(a[-1]).shape:   (20,100)
    #(a[-2].T).shape: (100,20)
    grads[-1] = a[-2].T.dot(delta) #grads[-1].shape: (100,10)
    print(f"len(a) is: {len(a)}")
    i = 1
    #for i in range(len(a)-2, 0, -1):
    print(f'  in loop i: {i}')
    #print("   delta.shape: "+str(delta.shape))
    print("   (weights[i].T).shape: " + str((weights[i].T).shape))
    delta = (a[i] > 0) * delta.dot(weights[i].T)
    print("   delta.shape: "+str(delta.shape)) #(20,100)
    grads[i-1] = a[i-1].T.dot(delta)
    print(f"   grads.shape: {grads.shape}")
    return grads / len(X)

trX, trY, teX, teY = mnist.load_data()
weights = [np.random.randn(*w) * 0.1 for w in [(784, 100), (100, 10)]]
num_epochs, batch_size, learn_rate = 30, 20, 0.1

for i in range(num_epochs):
    print(f'i is: {i}')
    for j in range(0, len(trX), batch_size):
        print(f'j is {j}')
        X, Y = trX[j:j+batch_size], trY[j:j+batch_size]
        print(f'X.shape: {X.shape}')
        print(f'Y.shape: {Y.shape}')
        gs  = grads(X, Y, weights)
        print(f'gs[0].shape is: {(gs[0]).shape}')
        print(f'gs[1].shape is: {(gs[1]).shape}')
        print(f'weights[0].shape is: {(weights[0]).shape}')
        print(f'weights[1].shape is: {(weights[1]).shape}')
        weights -= learn_rate * gs
    prediction = np.argmax(feed_forward(teX, weights)[-1], axis=1)
    print(f'Epoch: {i}, {np.mean(prediction == np.argmax(teY, axis=1))}')

File: mnist_in_30.jl
using MLDatasets
using LinearAlgebra
using Statistics
using PyCall
np = pyimport("numpy")
using Debugger
using ImageShow
       
function feed_forward(X, weights::Array{Array{Float32,2},1})
    a = [float(X)]
    for w in weights
        p = max.(w'*a[end],0)
        push!(a,p)
    end
    return a
end

#Calculate Gradients
function grads(X, Y::Array{Float32,2}, weights::Array{Array{Float32,2},1})::Array{Array{Float32,2},1}
    grads = similar(weights)
    a = feed_forward(X, weights)
    delta = a[end]' - Y # difference between expected (Y) and final layer output (a[end])
    grads[end] = a[end-1]*delta 
    i = 2
        dw = (delta * weights[i]')
        relu = (a[i] .> 0)
        delta = dw .* relu'
        grads[i-1] = a[i-1] * delta   #* delta
    return grads ./ size(X)[2]
end

#breakpoint(grads)

function to_one_hot(v,classes)
   out_mat = zeros(size(v)[1],classes)
   for i in 1:length(v)
     out_mat[i,v[i]+1] = 1.0
   end
   return out_mat
end

trX, trY = MNIST.traindata()
#TODO: need to one-hot-encode the trY   
trY = to_one_hot(trY,10)

teX, teY = MNIST.testdata()
t = size(trX)
trX = reshape(trX, (t[1]*t[2]), t[3])
t = size(teX)
teX = reshape(teX, (t[1]*t[2]), t[3])
 #teY = reshape(teY, (t[1]*t[2]), t[3])

trainset = (trX,trY)
testset  = (teX,teY)
trY = convert(Array{Float32,2}, trY)
teY = convert(Array{Float32,1}, teY)

## Helper funcs  for debugging ###
#
function predict(n, mx=teX, my=teY)
    ff = feed_forward(mx, weights)
    prediction = argmax(ff[end][:,n]) - 1
    return prediction
end

function comp_tr_img(n)
    imshow(Gray.(reshape(trX[:,n], (28,28))))
    p = predict(n,trX, trY)
    expected = argmax(trY[n, :])-1
    println("expected: $expected")
    return expected
end

#show an image of the expected output
function compare_img(n,mx,my)
    imshow(Gray.(reshape(mx[:,n], (28,28))))
    p = predict(n, mx, my)
    expected = my[n,:][1]
    println("expected: $expected")
    println("predicted: $p")
    if( p != expected )
        println("MISMATCH")
    else
        println("Correct!")
    end
    return expected
end

#all weights are 0.5:
#weights = [fill(0.5, w)  for w in [(784, 100), (100, 10)]]

#random weights:
weights = [randn(Float32, w) * 0.1 for w in [(784, 100), (100, 10)]]
weights = convert(Array{Array{Float32,2},1}, weights)
@show typeof(weights)

##load Python weights instead:
#weights = np.load("weights.npy", allow_pickle=true)
#weights = convert(Array{Array{Float32,2},1}, weights)
num_epochs, batch_size, learn_rate = 3, 20, Float32(0.1)


function run(num_epochs, trX, trY, teX, teY, batch_size, w::Array{Array{Float32,2},1}, learn_rate::Float32 )
    weights::Array{Array{Float32,2},1} = w
    @show typeof(weights)
    for i in 1:num_epochs
        print("epoch: $i")
        for j in 1:batch_size:size(trX)[2]
            @inbounds X, Y = trX[:,j:j+batch_size-1], trY[j:j+batch_size-1,:]
            gs::Array{Array{Float32,2},1} = grads(X, Y, weights)
            weights -= learn_rate .* gs
        end
        ff = feed_forward(teX, weights)
        prediction = [x[1] for x in (argmax(ff[end], dims=1))[1,:]]
        prediction = [i[1]-1 for i in prediction]
        equiv = (prediction .== teY) #this was possibly the problem
        println(" accuracy%: $(mean(equiv))")
    end
end

#@enter run(num_epochs, trX, trY, teX, teY, batch_size, weights, learn_rate )
@time run(num_epochs, trX, trY, teX, teY, batch_size, weights, learn_rate )